### PR TITLE
Use Base parallel precompilation to build stdlibs

### PR DESCRIPTION
**Pull request by IanButterworth**
Follow-on from https://github.com/JuliaLang/julia/pull/53403

This extends `Base.Precompilation.precompilepkgs` to take a list of configurations to precompile each package with, while parallelizing across all packages and configurations, and uses it to build the stdlib pkgimages.

It simplifies the stdlib pkgimage build process but is (currently) dependent on having an accurately resolved Manifest.toml (Project.toml included to make the manifest easier to make). Any new/removed stdlibs or changes their dependencies will require updating the Manifest.toml. It's a bit chicken and egg, but should be manageable with manual editing of the Manifest.toml.

In terms of speed improvement:
MacOS aarch64 CI runner 6m19s before, 5m19 with this

Note that CI builds will show the basic print with timing of each package, whereas local build will be the tidier fancy print without timings.

Todo
- [x] Figure out how to avoid printing `JULIA all-release` when re-build is a noop
```
julia % make
    JULIA all-release
julia % 
```

### CODE REVIEW

File: pkgimage.mk [hunk 1]
```diff
@@ -7,7 +7,7 @@ include $(JULIAHOME)/stdlib/stdlib.mk
 
 # set some influential environment variables
 export JULIA_DEPOT_PATH := $(build_prefix)/share/julia
-export JULIA_LOAD_PATH := @stdlib
+export JULIA_LOAD_PATH := @stdlib:$(JULIAHOME)/stdlib
```
Comment (vchuravy) [inline pkgimage.mk:10]: Why is this necessary?

File: pkgimage.mk [hunk 2]
```diff
@@ -24,129 +24,5 @@ $(JULIA_DEPOT_PATH):
...
-								  TOML Tar UUIDs p7zip_jll))
-
-# 7-depth packages
-$(eval $(call stdlib_builder,LazyArtifacts,Artifacts Pkg))
-$(eval $(call stdlib_builder,REPLExt,Pkg REPL))
-
-$(eval $(call stdlib_builder,SparseArrays,Libdl LinearAlgebra Random Serialization SuiteSparse_jll))
-$(eval $(call stdlib_builder,Statistics,LinearAlgebra SparseArrays))
+all-release all-debug:
+	@$(call PRINT_JULIA, $(call spawn,$(JULIA_EXECUTABLE)) --startup-file=no -e 'Base.Precompilation.precompilepkgs(;configs=[``=>Base.CacheFlags(), `--check-bounds=yes`=>Base.CacheFlags(true, 1, 1, true, 2)])')
```
Comment (vchuravy) [inline pkgimage.mk:28]: Should we make `CacheFlags` `@kwdef` and add a copy constructor for this? So that it is just `CacheFlags(CacheFlags(), check_bounds=2)` (or `=yes`

File: stdlib/Project.toml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+[deps]
```
Comment (vchuravy) [inline stdlib/Project.toml:1]: I think this is absolutely the right step. Thoughts for the future:

Could we get away with the current special way of managing versions in `stdlib` and instead use compat bounds? But that might require Julia to build Julia something that we have so far avoided. But we could totally use Julia 1.11 to build Julia 1.12

Should we have compat bounds here?

File: pkgimage.mk [hunk 4]
```diff
@@ -24,129 +24,5 @@ $(JULIA_DEPOT_PATH):
...
-								  Logging Markdown Printf REPL Random SHA Serialization\
-								  TOML Tar UUIDs p7zip_jll))
-
-# 7-depth packages
-$(eval $(call stdlib_builder,LazyArtifacts,Artifacts Pkg))
-$(eval $(call stdlib_builder,REPLExt,Pkg REPL))
-
-$(eval $(call stdlib_builder,SparseArrays,Libdl LinearAlgebra Random Serialization SuiteSparse_jll))
-$(eval $(call stdlib_builder,Statistics,LinearAlgebra SparseArrays))
+all-release all-debug:
```
Comment (vchuravy) [inline pkgimage.mk:27]: Things we are losing here right now.

`stdlib_builder` had a dependency on the source files, the Project.toml and the version file for the stdlib.

Similar to `SYSIMG_STDLIB_SRCS` we will need `STDLIB_SRCS` for the independent stdlibs and use that as a condition to this build step.

Also it's probably better to split this in to two and have a marker file instead of a phony target, so that make can skip running this step when all the preconditions are up to date.

File: pkgimage.mk [hunk 5]
```diff
@@ -24,129 +24,5 @@ $(JULIA_DEPOT_PATH):
 print-depot-path:
 	@$(call PRINT_JULIA, $(call spawn,$(JULIA_EXECUTABLE)) --startup-file=no -e '@show Base.DEPOT_PATH')
 
-all-release: $(addprefix cache-release-, $(STDLIBS))
-all-debug:   $(addprefix cache-debug-, $(STDLIBS))
-
-define stdlib_builder
-ifneq ($(filter $(1),$(INDEPENDENT_STDLIBS) $(PKG_EXTS)),)
-# Define target-specific export for `JULIA_CPU_TARGET`
```
Comment (vchuravy) [inline pkgimage.mk:32]: Is `JULIA_CPU_TARGET` handled currently?

File: pkgimage.mk (hunk 4 above)
Comment (vchuravy) [reply pkgimage.mk:27]: Another precondition needs to be `$(build_private_libdir)/sys.$(SHLIB_EXT)`

**Review by vchuravy — COMMENTED** [review]
This is fantastic! Unable to comment on the changes to `precompilation.jl`, but the rest looks great.

Need to adjust some of Makefile handling to ensure that we only run the step when needed and not all the time when calling make.

I wonder if we could parse the project.toml to have a source of truth for all the stdlibs instead of having a manual list (and so that the system complains if someone just updates the Project.toml but not the `INDEPENDENT_STDLIBS`

File: pkgimage.mk (hunk 1 above)
Comment (IanButterworth) [reply pkgimage.mk:10]: It doesn't know about the extensions (SparseArraysExt and REPLExt) if this isn't like this. Not sure I 100% understand, but this did fix that issue. @KristofferC sound reasonable?

Comment (KristofferC) [reply pkgimage.mk:10]: Hm, ideally it should find them even with only `@stdlib`. Not sure why this would fix it.

Comment (IanButterworth) [reply pkgimage.mk:10]: Probably about having the right active project? Note that the Project.toml and Manifest.toml sit in `$(JULIAHOME)/stdlib/`

Comment (vchuravy) [reply pkgimage.mk:10]: Should we instead set `JULIA_PROJECT`?

Comment (KristofferC) [reply pkgimage.mk:10]: > Probably about having the right active project? Note that the Project.toml and Manifest.toml sit in $(JULIAHOME)/stdlib/

Okay, in that case it makes sense I think.

Comment (IanButterworth) [reply pkgimage.mk:10]: Would that be different to `--project=$(JULIAHOME)/stdlib` which is what I had before that didn't work

File: pkgimage.mk (hunk 5 above)
Comment (IanButterworth) [reply pkgimage.mk:32]: How should it be?

File: pkgimage.mk (hunk 4 above)
Comment (IanButterworth) [reply pkgimage.mk:27]: Calling `precompilepkgs` is a noop if there's nothing to do. That's one of its strengths. I think we just want to call it every time? The only thing thats awkward is that currently make will print this for a noop
```
julia % make -j6
    JULIA all-release
julia % 
```

File: pkgimage.mk (hunk 2 above)
Comment (IanButterworth) [reply pkgimage.mk]: Done

File: base/precompilation.jl [hunk 6]
```diff
@@ -338,6 +338,9 @@ end
 const Config = Pair{Cmd, Base.CacheFlags}
 const PkgConfig = Tuple{Base.PkgId,Config}
 
+# TODO: Remove once Pkg moves to the renamed kwarg `flags_cacheflags` -> `configs`
+precompilepkgs(args...; flags_cacheflags, kwargs...) = precompilepkgs(args...; configs=flags_cacheflags, kwargs...)
```
Comment (KristofferC) [inline base/precompilation.jl:342]: I don't think this works since the positional arguments are the same? Probably need to handle it in the body of the original function?

Comment (IanButterworth) [reply base/precompilation.jl]: Ah ok. Fixed

**Comment by vchuravy** [comment]
@vtjnash could you take a quick look at the Makefile changes?

**Comment by IanButterworth** [comment]
Apart from the makefile review, this should be ready to merge

**Review by vchuravy — APPROVED** [review]

**Comment by JeffBezanson** [comment]
This is really nice, thanks!
